pkg/proxy/iptables/number_generated_rules_test.go

```diff
@@ -0,0 +1,245 @@
+// - ServiceType
+// - Number of Services
+// - Number of Endpoints per Service
+// This test will fail when the number of rules change, so the person
+// that is modifying the code can have feedback about the performance impact
+// on their changes. It also runs multiple number of rules test cases to check
+// if the number of rules grows linearly.
+func TestNumberIptablesRules(t *testing.T) {
+	testCases := []struct {
+		name                string
+		serviceType         v1.ServiceType
+		services            int
+		epPerService        int
+		expectedNatRules    int
+		expectedFilterRules int
+	}{
+		{
+			name:                "ClusterIP 500 Services 500 EndpointPerService",
+			serviceType:         v1.ServiceTypeClusterIP,
+			services:            500,
+			epPerService:        500,
+			expectedNatRules:    2505,
+			expectedFilterRules: 3,
+		},
+		{
+			name:                "LoadBalancer 500 Services 500 EndpointPerService",
+			serviceType:         v1.ServiceTypeLoadBalancer,
+			services:            500,
+			epPerService:        500,
+			expectedNatRules:    6005,
+			expectedFilterRules: 3,
+		},
+		{
+			name:                "NodePort 500 Services 500 EndpointPerService",
+			serviceType:         v1.ServiceTypeNodePort,
+			services:            500,
+			epPerService:        500,
+			expectedNatRules:    4005,
+			expectedFilterRules: 3,
+		},
+		{
+			name:                "ClusterIP 1000 Services 1000 EndpointPerService",
+			serviceType:         v1.ServiceTypeClusterIP,
+			services:            1000,
+			epPerService:        1000,
+			expectedNatRules:    5005,
+			expectedFilterRules: 3,
+		},
+		{
+			name:                "LoadBalancer 1000 Services 1000 EndpointPerService",
+			serviceType:         v1.ServiceTypeLoadBalancer,
+			services:            1000,
+			epPerService:        1000,
+			expectedNatRules:    12005,
+			expectedFilterRules: 3,
+		},
+		{
+			name:                "NodePort 1000 Services 1000 EndpointPerService",
+			serviceType:         v1.ServiceTypeNodePort,
+			services:            1000,
+			epPerService:        1000,
+			expectedNatRules:    8005,
+			expectedFilterRules: 3,
+		},
+		{
+			name:                "ClusterIP 5000 Services 5000 EndpointPerService",
+			serviceType:         v1.ServiceTypeClusterIP,
+			services:            5000,
+			epPerService:        5000,
+			expectedNatRules:    25005,
+			expectedFilterRules: 3,
+		},
+		{
+			name:                "LoadBalancer 5000 Services 5000 EndpointPerService",
+			serviceType:         v1.ServiceTypeLoadBalancer,
+			services:            5000,
+			epPerService:        5000,
+			expectedNatRules:    60005,
+			expectedFilterRules: 3,
+		},
+		{
+			name:                "NodePort 5000 Services 5000 EndpointPerService",
+			serviceType:         v1.ServiceTypeNodePort,
+			services:            5000,
+			epPerService:        5000,
+			expectedNatRules:    40005,
+			expectedFilterRules: 3,
+		},
+	}
+
+	for _, test := range testCases {
```

Comment:
What about ITP, ETP, affinity, lb source ranges, etc?  Those all change the results.

Reply:
this is just to have a ballpark estimation and a simple test to avoid regressions?
 do we really want to have all those numbers or is a nice to have?

Reply:
This is only partial confidence without the other cases.  It's just as likely that ETP/ITP logic breaks (more likely, in fact since that is under active development).  Affinity is less interesting but does change the result.

Reply:
I'm trying to have a way to verify that when we refactor the kube-proxy logic we really reduce the number of lines generated, right now we don't know it.

Our scale problems are related to the number of lines generated, those features are a subset of the rules we generate or will be a constant, right? we do verify the exact rules generated in the other tests

Reply:
We do verify elsewhere, so I'm OK here, I guess.

Reply:
https://github.com/kubernetes/kubernetes/blob/7ed2f1d94d694c6c4fdb4629638c38b1cbda7288/pkg/proxy/iptables/proxier_test.go#L3063-L3076